### PR TITLE
Log course sync timing

### PR DESCRIPTION
**Pull request by nwalters512**
This PR adds user-facing timing information for course syncs. This should help us identify what's slowing down syncs in the wild.

Screenshot 2024-09-06 at 09 47 32

### CODE REVIEW

File: apps/prairielearn/src/sync/syncFromDisk.ts [hunk 1]
```diff
@@ -37,35 +34,56 @@ export async function syncDiskToSqlWithLock(
...
+    );
+
+    await timed('Synced tags', () => syncTags.sync(courseId, courseData, questionIds));
+    await timed('Synced assessment sets', () => syncAssessmentSets.sync(courseId, courseData));
+    await timed('Synced assessment modules', () =>
+      syncAssessmentModules.sync(courseId, courseData),
+    );
+    await timed('Synced all assessments', async () => {
+      // Ensure that a single course with a ton of course instances can't
+      // monopolize the database connection pool.
```
Comment (nwalters512) [inline apps/prairielearn/src/sync/syncFromDisk.ts:73]: While testing all this, I realized that we would try to sync all course instances in parallel. Every call would queue up a DB query, so for courses with a lot of course instances (we have some with almost 20 in practice), a single sync would tie up the DB connection pool for many seconds. I think this explains the periodic "freezes" I've seen on instructor server routes.

Limiting concurrency has two benefits: it leaves other clients in the pool, and it also allows other queries to join the queue, making everything more fair.

File: apps/prairielearn/src/sync/syncFromDisk.ts [hunk 2]
```diff
@@ -37,35 +34,56 @@ export async function syncDiskToSqlWithLock(
   courseDir: string,
   logger: Logger,
 ): Promise<SyncResults> {
+  async function timed<T>(label: string, fn: () => Promise<T>): Promise<T> {
```
Comment (mwest1066) [inline apps/prairielearn/src/sync/syncFromDisk.ts:37]: Consider lifting this into a library function that takes a longer and returns the function. I feel like we should be using this generically in more server jobs.

File: apps/prairielearn/src/sync/syncFromDisk.ts (hunk 1 above)
Comment (mwest1066) [reply apps/prairielearn/src/sync/syncFromDisk.ts:73]: Great catch, we never want unlimited concurrency, I think.

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/sync/syncFromDisk.ts (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/sync/syncFromDisk.ts:37]: I'll do this in a future PR when we start using this in another place.

**Pull request merged**